src/main/kotlin/me/owdding/lib/waypoints/MeowddingWaypointHandler.kt

```diff
@@ -0,0 +1,110 @@
+
+    private var _waypoints: MutableList<MeowddingWaypoint> = mutableListOf()
+    val waypoints: List<MeowddingWaypoint> get() = _waypoints
+
+    fun addWaypoint(waypoint: MeowddingWaypoint) {
+        _waypoints.add(waypoint)
+
+        if (waypoint.inLocatorBar && !McVersionGroup.MC_1_21_5.isActive) {
+            waypoint.minecraftWaypoint = MinecraftWaypointHandler.addWaypoint(waypoint)
+        }
+    }
+
+    fun removeWaypoint(uuid: UUID) {
+        _waypoints.find { it.uuid == uuid }?.let(::removeWaypoint)
+    }
+
+    fun removeWaypoint(waypoint: MeowddingWaypoint) {
+        _waypoints.remove(waypoint)
+        if (waypoint.inLocatorBar && !McVersionGroup.MC_1_21_5.isActive) {
+            MinecraftWaypointHandler.removeWaypoint(waypoint)
+        }
+    }
+
+    @Subscription
+    fun onCommand(event: RegisterCommandsEvent) {
+        event.register("meowdding waypoint") {
+            then("add") {
+                thenCallback("coords", BlockPosArgument.blockPos()) {
+                    val pos = this.getArgument("coords", Coordinates::class.java).getBlockPos(this.source.toCommandSourceStack())
+                    MeowddingWaypoint(pos) {
+                        withName("Waypoint $uuid")
+                        withRandomColor()
+                        withAllRenderTypes()
+                        inLocatorBar()
+                    }
+                }
+                callback {
+                    MeowddingWaypoint(McPlayer.position!!) {
+                        withName("Waypoint $uuid")
+                        withRandomColor()
+                        withAllRenderTypes()
+                        inLocatorBar()
+                    }
+                }
+            }
+            then("remove") {
+                thenCallback("all") {
+                    _waypoints.forEach(::removeWaypoint)
+                }
+                thenCallback("nearest") {
+                    val position = McPlayer.position ?: return@thenCallback
+                    _waypoints.minByOrNull { it.distanceToSqr(position) }?.let(::removeWaypoint)
+                }
+                thenCallback("uuid uuid", StringArgumentType.string(), MeowddingSuggestionProviders.iterable(_waypoints) { it.uuid.toString() }) {
+                    val uuid = this.getArgument("uuid", String::class.java)
+                    _waypoints.find { it.uuid.toString() == uuid }?.let(::removeWaypoint)
+                }
+                thenCallback("name name", StringArgumentType.greedyString(), MeowddingSuggestionProviders.iterable(_waypoints) { it.name.stripped }) {
+                    val name = this.getArgument("name", String::class.java)
+                    _waypoints.find { it.name.stripped.equals(name, ignoreCase = true) }?.let(::removeWaypoint)
+                }
+            }
+        }
+    }
+
+    @Subscription(ServerChangeEvent::class, ServerDisconnectEvent::class)
+    fun clearWaypoints() {
+        _waypoints.clear()
+    }
```

Comment:
personally i'd do this
```suggestion
    fun onLeave()  = _waypoints.clear()
```